File: 03-1-login_kt2/app/src/main/java/com/example/login_kt2/MainViewModel.kt
package com.example.login_kt2

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.login_kt2.repository.LoginTryRepository
import com.example.login_kt2.repository.UserRepository
import com.example.login_kt2.repository.room.LoginTry
import kotlinx.coroutines.launch
import java.util.Date

class MainViewModel(val loginTryRepository: LoginTryRepository) : ViewModel() {

    var username by mutableStateOf("")
    var usernameError by mutableStateOf(false)
    var usernameErrorMessage by mutableStateOf("")
    var password by mutableStateOf("")
    var passwordError by mutableStateOf(false)
    var passwordErrorMessage by mutableStateOf("")
    var isLoading by mutableStateOf(false)
    var isLoginSuccessful by mutableStateOf(false)

    val userRepository = UserRepository()

    fun performLogin() {
        usernameError = false
        passwordError = false

        username = username.trim()
        if (username == "") {
            usernameError = true
            usernameErrorMessage = "empty"
            return
        }

        if (password == "") {
            passwordError = true
            passwordErrorMessage = "empty"
            return
        }

        viewModelScope.launch {
            isLoading = true
            var loginTry = LoginTry(
                username = username,
                passwordLength = password.length,
                whenTried = Date().toString(),
                wasSuccessful = false
            )
            val status = userRepository.login(username, password)
            isLoading = false
            Log.i("MainViewModel", "name: $username, password: $password, status: $status")
            when(status) {
                "success" -> {
                    loginTry = loginTry.copy(wasSuccessful = true)
                    Log.i("MainViewModel", "moving to main screen")
                    isLoginSuccessful = true
                }
                "wrong_username" -> {
                    usernameError = true
                    usernameErrorMessage = "wrong user name"
                }
                "wrong_password" -> {
                    passwordError = true
                    passwordErrorMessage = "wrong password"
                }
                else -> {
                    usernameError = true
                    usernameErrorMessage = "Error: $status"
                }
            }
            Log.i("MainViewModel", "insert login try: $loginTry")
            loginTryRepository.insert(loginTry)
        }
    }

    fun clearLogin() {
        username = ""
        password = ""
        usernameError = false
        passwordError = false
    }

    fun notifyTransition() {
        isLoginSuccessful = false
        clearLogin()
    }
}

File: 03-1-login_fl/lib/main_view_model.dart
import 'package:flutter/material.dart';
import 'package:login_fl/repository/user_repository.dart';

class MainViewModel extends ChangeNotifier {

  final UserRepository userRepository;

  final TextEditingController usernameController = TextEditingController();
  final TextEditingController passwordController = TextEditingController();

  String? _usernameError;
  String? _passwordError;

  bool _isLoading = false;

  MainViewModel(this.userRepository);

  String? get usernameError => _usernameError;
  String? get passwordError => _passwordError;
  bool get isLoading => _isLoading;

  Future<void> performLogin(void Function() onSuccess) async {
    print("Login button clicked");
    print("Username: ${usernameController.text}, password: ${passwordController
        .text}");

    _usernameError = null;
    _passwordError = null;

    if (usernameController.text == "") {
      _usernameError = "empty";
      notifyListeners();
      return;
    }

    if (passwordController.text == "") {
      _passwordError = "empty";
      notifyListeners();
      return;
    }

    _isLoading = true;
    notifyListeners();

    var status = await userRepository.login(
        usernameController.text, passwordController.text);

    if (status == "wrong_username") {
      _usernameError = "wrong user name";
    } else if (status == "wrong_password") {
      _passwordError = "wrong password";
    } else {
      _usernameError = "Error: $status";
    }
    _isLoading = false;
    notifyListeners();

    if (status == "success") {
      print("MainViewModel: moving to main screen");
      onSuccess();
    }
  }

  void clearLogin() {
    print("Clear button clicked");
    usernameController.clear();
    passwordController.clear();
    _usernameError = null;
    _passwordError = null;
    notifyListeners();
  }
}